Increase network packet size to for IPv6 and for loopback

Rogier-5 opened the pull request:
Edit: These patches now change the network packet size for IPv6 and loopback only.
They may be increased for other cases as well, but it remains to be seen how that
can best be done.

------------------------------------------------------------------------

A size of 512 is way too small for normal operation, and will make
inefficient use of CPU, network bandwidth and network processing power.

The value of 4000 significantly reduces the overhead, makes one UDP packet
fit in 3 ethernet packets even if the MTU is a bit less than 1500, and it
is not too large, so that the impact of packet loss is still limited.

For peers that don't support protocol v29, the maximum supported packet
size is 1500, so the chosen maximum is set a bit lower in case
some kind of encapsulation is used.

I tested all combinations of old & new client and server, and they
work fine.

est31 commented:
:-1: this is more complicated than just setting a number. It won't work without a sophisticated MTU finding algorithm, otherwise connections will fail on networks with small MTUs.

nerzhul commented:
@Rogier-5 interesting idea to reduce latency on heavy servers sending big string data, but i think using 1000 is sufficient, 500 is ridiculous.

Rogier-5 commented:
> interesting idea to reduce latency on heavy servers sending big string data, but i think using 1000 is sufficient

@est31 does have a point, that a packet size greater than 576 (for IPv4, including UDP/IP headers) is not *guaranteed* to be deliverable. In practise, most people playing minetest will communicate via modern equipment, which I assume will have a sufficiently large MTU, or else will fragment packets. But there is no guarantee. In addition, non-source packet fragmentation has issues, and should be used as a workaround only.

To avoid the possibly controversial part for the moment, I modified the code to do the following:
- Use a packet size of 1200 for IPv6 (IPv6 should support 1280 bytes without fragmentation, and 1500 with possible fragmentation)
- Use a packet size of 65000 for the loopback network.

I removed the general packet size increase for the moment.
As far as I understand, linux can do path MTU discovery for UDP. I don't know about Windows. Maybe that that feature can be enabled and made to work later.

est31 commented:
Removing my :-1: . The new approach looks sensible.

ShadowNinja commented on src/socket.cpp:
```diff
@@ -272,6 +283,24 @@ bool Address::isZero() const
+		u32 addr = ntohl(m_address.ipv4.sin_addr.s_addr);
+		return (addr & 0xff000000) == 0x7f000000;
+	}
+}
```
There are library-defined constants for this... https://github.com/minetest/minetest/pull/2604/files#diff-2fad06c037d3ec2891fb90f6e1325647R200

numberZero replied:
But this seems to be the correct way, while the code you reference just compares with a constant, probably `127.0.0.1` (don’t forget that e.g. `127.3.4.5` is loopback too)

sapier commented:
Any device supporting IPv6 **has to** support MTU >=1500 bytes thus increasing to 1200 as soon as ipv6 is used will not cause any issues.

rubenwardy commented:
This seems useful. Has anyone tested it thoroughly?

nerzhul commented:
4000 is not a good value, but we can increase packet size to 768 if needed, 512 MTU is very very very rare, it's on very old ethernet networks, not very suitable for minetest remote play

paramat commented:
Needs adoption, author disappeared.

paramat commented:
Neglected so closing.

paramat commented:
Waiting for adoption, rebase and testing.

paramat commented:
No adoption interest, closing for now, but tagging @nerzhul

paramat commented:
@nerzhul are you interested in adopting this?

paramat commented:
Nerzhul expressed an interest in adopting this sometime.
